api/api-agents-v4/src/main/java/com/thoughtworks/go/apiv4/agents/AgentsControllerV4.java

```diff
@@ -0,0 +1,208 @@
+            return writerForTopLevelObject(request, response, outputWriter -> AgentRepresenter.toJSON(outputWriter, updatedAgentInstance, environmentConfigService.environmentsFor(uuid), securityService, currentUsername()));
+        }
+
+        return renderHTTPOperationResult(result, request, response);
+    }
+
+    public String bulkUpdate(Request request, Response response) throws IOException {
+        final AgentBulkUpdateRequest bulkUpdateRequest = AgentBulkUpdateRequestRepresenter.fromJSON(request.body());
+
+        final HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        agentService.bulkUpdateAgentAttributes(currentUsername(),
+                result,
+                bulkUpdateRequest.getUuids(),
+                bulkUpdateRequest.getOperations().getResources().toAdd(),
+                bulkUpdateRequest.getOperations().getResources().toRemove(),
+                bulkUpdateRequest.getOperations().getEnvironments().toAdd(),
+                bulkUpdateRequest.getOperations().getEnvironments().toRemove(),
+                bulkUpdateRequest.getAgentConfigState()
+        );
+
+        return renderHTTPOperationResult(result, request, response);
+    }
+
+    public String deleteAgent(Request request, Response response) throws IOException {
+        final HttpOperationResult result = new HttpOperationResult();
+        agentService.deleteAgents(currentUsername(), result, singletonList(request.params("uuid")));
+        return renderHTTPOperationResult(result, request, response);
+    }
+
+    public String bulkDeleteAgents(Request request, Response response) throws IOException {
+        final JsonReader reader = GsonTransformer.getInstance().jsonReaderFrom(request.body());
+        final List<String> uuids = toList(reader.optJsonArray("uuids").orElse(new JsonArray()));
+
+        final HttpOperationResult result = new HttpOperationResult();
+        agentService.deleteAgents(currentUsername(), result, uuids);
+
+        return renderHTTPOperationResult(result, request, response);
+    }
+
+    @Override
+    public String etagFor(AgentInstance entityFromServer) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public AgentInstance doFetchEntityFromConfig(String uuid) {
+        final AgentInstance agentInstance = agentService.findAgent(uuid);
+
+        return agentInstance instanceof NullAgentInstance ? null : agentInstance;
+    }
+
+    @Override
+    public AgentInstance buildEntityFromRequestBody(Request req) {
+        return null;
+    }
+
+    @Override
+    public String jsonize(Request req, AgentInstance o) {
+        return null; // to be implemented
+    }
+
+    @Override
+    public Consumer<OutputWriter> jsonWriter(AgentInstance agentConfigs) {
+        return null;
+    }
+
+    private void checkSecurityOr403(Request request, Response response) {
+        if (Arrays.asList("PATCH", "PUT", "DELETE").contains(request.requestMethod().toUpperCase())) {
```

Comment:
Can we invert this, so GET and HEAD are allowed, everything else is denied, prevents accidentally opening up anything else?

Reply:
Sure, I will do it